lib/task.js

```diff
@@ -128,12 +128,55 @@ Task.prototype.chain = function _chain(f) {
  * @summary @Task[α, (β → γ)] => Task[α, β] → Task[α, γ]
  */
 
-Task.prototype.ap = function _ap(f2) {
-  return this.chain(function(f) {
-    return f2.map(f);
-  });
+Task.prototype.ap = function _ap(that) {
+  var forkThis = this.fork;
+  var forkThat = that.fork;
+  var cleanupThis = this.cleanup;
+  var cleanupThat = that.cleanup;
+  var func = null;
+  var val = null;
+
+  function cleanupBoth(state) {
+    cleanupThis(state[0]);
+    cleanupThat(state[1]);
+  }
+
+  return new Task(function(reject, resolve) {
+    var allState;
+    var thisState = forkThis(guardFn(reject, true), guardFn(resolve));
+    var thatState = forkThat(guardVal(reject, true), guardVal(resolve));
+
+    return allState = [thisState, thatState];
+
+    function guardFn(f, isFail) {
+      return function(x) {
+        if(isFail) return f(x);
+
+        if (val) {
+          delayed(function(){ cleanupBoth(allState) });
+          return f(func(x));
```

Comment:
@rjmk you are right. if `guardVal` is called, and some truthy value is stored in `val` and then `guardFn` is called `func` will be undefined.
also if `that` task will resolve to undefined or some falsy value and then `this` is resolved, `val` will be falsy so no one will call `func`